### ISSUE TITLE
pruner: filesystems with no correspondence on sink should not be pruned

### ISSUE THREAD

**Issue by JMoVS** [zrepl/zrepl#126]
Hi,

in my tests, I have now run into a strange problem where zrepl first complained this:

```
2019-03-02T11:39:54+01:00 [ERRO][backup_system][pruning]: cannot get replication cursor prune_side="receiver" fs="zfsintern/Users" err="endpoint does not allow access to filesystem zfsintern/Users" invocation="1"
2019-03-02T11:39:54+01:00 [ERRO][backup_system][pruning]: entering error state after error invocation="1" prune_side="receiver" err="endpoint does not allow access to filesystem zfsintern/Users" state="ErrPerm"
```

this was the config at the time:

```
jobs:
  - type: sink
    name: "local_sink"
    root_fs: "zfssnaps/encRoot"
    serve:
      type: local
      listener_name: localsink

  - type: push
    name: "backup_system"
    connect:
      type: local
      listener_name: localsink
      client_identity: SergeantPepperSenior
    filesystems: {
      "zfsintern/Users/justin/Cloudy<": true,
      "zweischock/temp<": true
    }
    snapshotting:
      type: periodic
      interval: 10m
      prefix: zrepl_
    pruning:
      keep_sender:
      - type: last_n
        count: 10
      keep_receiver:
      - type: grid
        grid: 1x1h(keep=all) | 24x1h(keep=1) | 35x1d(keep=1) | 6x30d(keep=1) | 24x30d(keep=1) | 6x90d(keep=1)
        regex: "zrepl_.*"

```

and this is the output of zfs list

```
zfssnaps                                                                             1,03T   476G  2,55M  none
zfssnaps/encRoot                                                                     41,7G   476G   152K  none
zfssnaps/encRoot/SergeantPepperSenior                                                41,7G   476G   152K  none
zfssnaps/encRoot/SergeantPepperSenior/zfsintern                                      36,8G   476G   152K  none
zfssnaps/encRoot/SergeantPepperSenior/zfsintern/Users                                4,10G   476G   184K  none
zfssnaps/encRoot/SergeantPepperSenior/zfsintern/Users/justin                         4,10G   476G   184K  none
zfssnaps/encRoot/SergeantPepperSenior/zfsintern/Users/justin/Cloudy                  4,10G   476G   101M  none
zfssnaps/encRoot/SergeantPepperSenior/zfsintern/Users/justin/Cloudy/dataset1  1,06G   476G  1,03G  none
zfssnaps/encRoot/SergeantPepperSenior/zfsintern/Users/justin/Cloudy/dataset2      1,43G   476G  1,43G  none
zfssnaps/encRoot/SergeantPepperSenior/zfsintern/Users/justin/Cloudy/dataset3          1014M   476G   897M  none
zfssnaps/encRoot/SergeantPepperSenior/zfsintern/Users/justin/Cloudy/dataset4              247M   476G   247M  none
zfssnaps/encRoot/SergeantPepperSenior/zfsintern/Users/justin/Cloudy/dataset5
```

zrepl is run with `sudo`, so there shouldn't be any access problems on a local job

I now *removed* all references to zfsintern from the config but it somehow still wants to access it? Where does zrepl keep state? This is the new config but the error is still the same?

```
jobs:
  - type: sink
    name: "local_sink"
    root_fs: "zfssnaps/encRoot"
    serve:
      type: local
      listener_name: localsink

  - type: push
    name: "backup_system"
    connect:
      type: local
      listener_name: localsink
      client_identity: SergeantPepperSenior
    filesystems: {
      "zweischock/temp": true
    }
    snapshotting:
      type: periodic
      interval: 1m
      prefix: zrepl_
    pruning:
      keep_sender:
      - type: last_n
        count: 10
      keep_receiver:
      - type: grid
        grid: 1x1h(keep=all) | 24x1h(keep=1) | 35x1d(keep=1) | 6x30d(keep=1) | 24x30d(keep=1) | 6x90d(keep=1)
        regex: "zrepl_.*"

```

where does it **now**, without any mention of zfsintern/Users get the idea that it should try to access that? I can access the filesystems myself - this is rather weird

**Comment by problame** [comment]
It does not keep state apart from the config file and implicit state in that the sink's `root_fs` sub-datasets are _all_ considered during replication.

It seems to work fine with regards to replication since the error log is from the pruner.
The pruner iteratres over the pruning target's filesystems (i.e. the `prune_side="receiver"` in your case)  and searches for the replication cursor on the source (the push job in your case): https://github.com/zrepl/zrepl/blob/problame/replication_refactor/daemon/pruner/pruner.go#L383 and https://github.com/zrepl/zrepl/blob/problame/replication_refactor/daemon/pruner/pruner.go#L417

** The replication cursor RPC fails because the receiver didn't notice the deletion / "removal from the filesystem acl". The source is thus behaving exactly the way it should, it's just that the pruner tries to prune filesystems that no longer have correspondence on the source.

This is an aspect of #70 . Will change the title and add it to #70 . Likely that this will be skipped for 0.1 though.

**Comment by jbliesener** [comment]
I've had the same error here when I renamed/moved some filesystems on the receiver and adjusted the job configuration accordingly (Ex: zfs rename zpool/data/zrepl_target_1 zpool/data/zrepl_target_2 and change the configuration files as well).

Turned out that the problem was that the zrepl:placeholder property on the destination filesystem did not match anymore and therefore, although all data was correctly replicated with all existing snapshots maintained, the pruner was unable to handle the destination filesystem.

Here's how I was able to fix it:
- `zfs rename zpool/data/zrepl_target_2 zpool/data/zrepl_target_1` (move back to the old location)
- do **not** change the config files and start a replication to the new location (`zpool/data/zrepl_target_2`)
- while the replication is running: `zfs list -o name,zrepl:placeholder`
- abort ongoing replication, we only need the correct value for `zrepl:placeholder`
- `zfs destroy -r zpool/data/zrepl_target_2`
- `zfs set zpool/data/zrepl_target_1/... <values obtained from zfs list>` (make sure to use inheritance)
- `zfs rename zpool/data/zrepl_target_1 zpool/data/zrepl_target_2` (you now have your previous data with all history and correct zrepl:placeholder properties where you want it)

**Comment by problame** [comment]
I'm not totally sure whether the described problems are related, but yes, the `zrepl:placeholder` property must be considered, too, with regards to renames.

**Comment by jbliesener** [comment]
Maybe if you could just print out the expected value for zrepl:placeholder in the logs, it would make things clearer to the user? I wasn't able to correctly calculate this hash value, as I didn't find how exactly it is generated in your source code.

**Comment by problame** [comment]
The idea was to have a `zrepl tool placeholder [mark|clear] zroot/path/to/placeholder/fs` subcommand that would mark|clear the filesystem as placeholder.
However, I would much prefer an automatic solution for the renaming feature.

**Comment by problame** [comment]
Oh, nvmd, there already is the `zrepl test placeholder` subcommand :D It just needs a `fix` variant.

**Comment by JMoVS** [comment]
oh yeah, that's a good hint @jbliesener ! I am moving my existing backups to an encrypted root on my backup pool and did some partial renaming etc, I will try inheriting the placeholder, didn't think of that. What is the role of the placeholder?

**Comment by problame** [comment]
On the sink, we always have `${root_fs}/${client_identity}/${sender_side_fs}`.
Now the source could only allow `zroot/foo/bar`, but we need the zfs datasets `${root_fs}/${client_identity}/{zroot,zroot/foo}` on the sink-side to receive into `${root_fs}/${client_identity}/zroot/foo/bar`.

This should be documented here: in https://zrepl.github.io/configuration/jobs.html?highlight=root_fs#how-the-passive-side-works
Should also mention the `zrepl test placeholder` subcommand.

Will convert this to a separate issue.

**Comment by JMoVS** [comment]
> Oh, nvmd, there already is the `zrepl test placeholder` subcommand :D It just needs a `fix` variant.

that doesn't seem to be in the rc3 version, right? `sudo zrepl test placeholder` returns un unknown command error to me

**Comment by problame** [comment]
Yup, landed in 2db3977408b906a0c7dc1bffc6b47d67397f0d6f which was after rc3.

**Comment by JMoVS** [comment]
I just ran the current master and it reported "skipped: filesystem has no correspondence on sender" - which is good!
